modules/@apostrophecms/page/index.js

```diff
@@ -976,6 +964,90 @@ database.`);
           }
         }
       },
+      // A method to return a target page object based on a passed `targetId`
+      // value. `position` is used to prevent attempts to move after the trash
+      // "page."
+      async getTarget(req, targetId, position) {
+        const criteria = (targetId === '_home') ? {
+          level: 0
+        } : (targetId === '_trash') ? {
+          level: 1,
+          trash: true
+        } : {
+          _id: targetId
+        };
+        const target = await self.find(req, criteria).permission(false).trash(null).published(null).areas(false).ancestors(_.assign({
+          depth: 1,
+          trash: null,
+          published: null,
+          areas: false,
+          permission: false
+        }, options.builders || {})).children({
+          depth: 1,
+          trash: null,
+          published: null,
+          areas: false,
+          permission: false
+        }).applyBuilders(options.builders || {}).toObject();
+        if (!target) {
+          throw self.apos.error('notfound');
+        }
+        if (target.type === '@apostrophecms/trash' && target.level === 1 && position === 'after') {
+          throw self.apos.error('invalid');
+        }
+        return target;
+      },
+      // A method to support numeric positions while moving pages within the
+      // page tree. If a numeric position is submitted, this method assumes
+      // that the `targetId` is meant to be the moved page's parent. It will
+      // return a new `targetId` for a sibling page to use the 'before' position
+      // if applicable.
+      async getTargetIdAndPosition(req, pageId, targetId, position) {
+        targetId = self.apos.launder.id(targetId);
+
+        position = self.apos.launder.string(position);
+
+        if (isNaN(parseInt(position))) {
+          // Return an already-valid position or a potentially invalid, but
+          // non-numeric position to be evaluated in `self.move`.
+          return {
+            targetId,
+            position
+          };
+        }
+
+        // The position is a number, so we're converting it to one of the
+        // acceptable string values and treating the `target` as the
+        // moved pages's parent.
+        const target = await self.getTarget(req, targetId, position);
+
+        position = parseInt(position);
+        // Get the index of the moving page within the target's children.
+        const childIndex = target._children.findIndex(child => {
+          return child._id === pageId;
+        });
+
+        if (position === 0 || target._children.length === 0) {
+          position = 'firstChild';
+        } else if (childIndex > -1 && position >= (target._children.length - 1)) {
+          position = 'lastChild';
+        } else {
+          if (childIndex === -1 && position >= (target._children.length)) {
+            position = 'lastChild';
```

Comment:
Two `lastChild` checks since it's different if the thing is moving from within the child group.